src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -26,114 +26,198 @@
-            }
-            i++;
-
-            try {
-                pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(field));
-            } catch (ParseException e) {
-                HttpHeaders headers = new HttpHeaders();
-                headers.add("errors", "date " + field + " not valid");
-                return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+            settingResponse = setDateOfBirth(pet, fields[1]);
+            if (!settingResponse.isSuccessful) {
+                return errorResponse(settingResponse.errorMessage);
             }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
+            settingResponse = setPetType(pet, fields[2]);
+            if (!settingResponse.isSuccessful) {
+                return errorResponse(settingResponse.errorMessage);
             }
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
+            settingResponse = setOwner(pet, fields[3]);
+            if (!settingResponse.isSuccessful) {
+                return errorResponse(settingResponse.errorMessage);
             }
 
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                if (matchingOwners.size() > 1) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not unique");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                pet.setOwner(matchingOwners.iterator().next());
+            String action = fields[4];
+            if (action.equalsIgnoreCase("add")) {
+                petsToAdd.add(pet);
+            } else if (action.equalsIgnoreCase("delete")) {
+                petsToDelete.add(pet);
+            } else {
+                return errorResponse("invalid action");
             }
+        }
+
+        for (Pet p : petsToAdd) {
```

Comment:
nice to also see such a solution ... you are the first ones so far that moved this functionality to end if all pets can be parsed...
it can get a bit tricky then to keep the same order of execution as intended by the original csv